### DISCUSSION TITLE
Is there an easy way to change a document's ID?

### DISCUSSION THREAD

**Discussion by ItsRobbAllen** [ravendb/ravendb#18039] [Q&A] [answered]
Let's say I have a document called `Reports/1-A` - is there any way to rename it to `Reports/1-A/processed` or `Reports/1-A/Errored` and keep things like metadata, counters, & attachments?

**Comment by ayende** [comment]
No, there is no such facilities.
Why do you want to _change_ the document id versus setting a status field?

**Comment by ItsRobbAllen** [comment]
Ah, because I'm thinking about it wrong. It's actually "How do I clone a document and give the clone a different Id?" which I can do with a PatchCommand (although attachments & counters won't automatically come over, but I don't have a use case for that right now so I won't worry about it).

The actual use case here is versioning mixed in with event sourcing & snapshots. We'll take a projection of events, save out a 'snapshot' of the event stream at that point in time, add more events, then eventually `publish` a version of this timeline. So we'll have `Timelines/32-C` as the _current state_ and `Timelines/32-C/V/1` as version 1, etc.

I was thinking that I could take the existing document, change its Id, then re-project the current state, but that's backwards.

Thanks!

**Reply by ayende** [reply, marked as answer]
For that scenario, isn't this when you use revisions?
You can either create a revision whenever you have a change, or "publish" a revision (which creates an immutable copy)

**Reply by ItsRobbAllen** [reply]
BECAUSE I DIDN'T REALIZE YOU COULD **MANUALLY** CREATE REVISIONS!!!

Using event sourcing, we'd end up creating thousands of revisions for no good reason (the events themselves tell me the history) so I didn't even _think_ about them in this capacity since I didn't know that was an option.

Holy SMOKES that's going to make this _so much easier_.

I've been using Raven since V2 and every time I turn around I keep finding new ways that you guys have already solved my problems for me.

**Reply by ayende** [reply]
Would you mind writing a blog post about this? That feature was specifically meant for this scenario, and we would love to make it more public

**Reply by ppekrol** [reply]
@ItsRobbAllen Could we use your last sentence as a quote on our website?

**Reply by ItsRobbAllen** [reply]
Yes to both questions :) 

However, for the blog post, let me get V1 out the door first. While we're using the event sourcing in another area, this one will be a little more involved & I'd like to have it released before I write up much about it.
